### PR TITLE
filter: set additional termination details and response flag for auth_digest_no_match in client_ssl_auth

### PR DESCRIPTION
**Pull request by agrawroh**
### Background

Currently, `client_ssl_auth` does not set any response flags or any other additional details while terminating the condition due to `auth_digest_no_match` error which occurs when the SSL certificate coming from the client is not in the allowed-list of principals.

**Motivation:** [See This](https://github.com/envoyproxy/envoy/issues/17847)

### Changes

This PR adds the response flag of `UAEX` and also sets `auth_digest_no_match` in the additional response details when the above mentioned conditions occur. 

---
**Commit Message:** set additional termination details and response flag for auth_digest_no_match in client_ssl_auth
**Additional Description:** See Background Section.
**Risk Level:** Low
**Testing:** Unit Tests
**Docs Changes:** N/A
**Release Notes:** Added
**Platform Specific Features:** N/A

**Signed-off-by:** Rohit Agrawal <rohit.agrawal@databricks.com>

### CODE REVIEW

File: test/extensions/filters/network/client_ssl_auth/client_ssl_auth_test.cc [hunk 1]
```diff
@@ -163,6 +163,10 @@ TEST_F(ClientSslAuthFilterTest, Ssl) {
       std::make_shared<Network::Address::Ipv4Instance>("192.168.1.1"));
   std::string expected_sha_1("digest");
   EXPECT_CALL(*ssl_, sha256PeerCertificateDigest()).WillOnce(ReturnRef(expected_sha_1));
+  EXPECT_CALL(filter_callbacks_.connection_.stream_info_,
+              setResponseFlag(StreamInfo::ResponseFlag::UpstreamProtocolError));
+  EXPECT_CALL(filter_callbacks_.connection_.stream_info_,
+              setResponseCodeDetails("auth_digest_no_match"));
   EXPECT_CALL(filter_callbacks_.connection_, close(Network::ConnectionCloseType::NoFlush));
```
Comment (agrawroh) [inline test/extensions/filters/network/client_ssl_auth/client_ssl_auth_test.cc:170]: @ggreenway `NoFlush` should still be okay right or do we need to change it to `FlushWrite`?

File: source/extensions/filters/network/client_ssl_auth/client_ssl_auth.cc [hunk 2]
```diff
@@ -22,6 +22,12 @@ namespace Extensions {
 namespace NetworkFilters {
 namespace ClientSslAuth {
 
+struct RcDetailsValues {
+  // SSL certificate is not in the allowed principals list.
+  const std::string AuthDigestNoMatch = "auth_digest_no_match";
+};
+using RcDetails = ConstSingleton<RcDetailsValues>;
+
```
Comment (wbpcode) [inline source/extensions/filters/network/client_ssl_auth/client_ssl_auth.cc:30]: May be simple `constexpr absl::string_view` would be better?

Comment (agrawroh) [reply source/extensions/filters/network/client_ssl_auth/client_ssl_auth.cc:30]: I just followed what we are doing in other filters.

https://github.com/envoyproxy/envoy/blob/9dd8ae3c69a9917d97a886ed03e1c010dcd9b098/source/extensions/filters/http/csrf/csrf_filter.cc#L21

https://github.com/envoyproxy/envoy/blob/5963beae8842982803af1bef04fb5a2a0893c613/source/extensions/filters/http/ratelimit/ratelimit.cc#L21

https://github.com/envoyproxy/envoy/blob/88705d72f063b7c75b18b452a4bd9af406611faf/source/extensions/filters/http/ext_authz/ext_authz.cc#L16

https://github.com/envoyproxy/envoy/blob/main/source/extensions/filters/http/fault/fault_filter.cc#L30

Should we keep it consistent instead or you prefer doing it this way here?

Comment (ggreenway) [reply source/extensions/filters/network/client_ssl_auth/client_ssl_auth.cc:30]: There's samples both ways. See `extensions/common/wasm/context.cc` `CloseStreamResponseDetails`. I think the constexpr string_view is simpler; please switch it to that.

File: test/extensions/filters/network/client_ssl_auth/client_ssl_auth_test.cc (hunk 1 above)
Comment (ggreenway) [reply test/extensions/filters/network/client_ssl_auth/client_ssl_auth_test.cc:170]: `NoFlush` should be fine. Are you seeing an issue with it?

**Review by ggreenway — CHANGES_REQUESTED** [review]
Looks good overall; thanks for improving this!

/wait

File: source/extensions/filters/network/client_ssl_auth/client_ssl_auth.cc (hunk 2 above)
Comment (agrawroh) [reply source/extensions/filters/network/client_ssl_auth/client_ssl_auth.cc:30]: Changed to `constexpr absl::string_view` as suggested.

File: test/extensions/filters/network/client_ssl_auth/client_ssl_auth_test.cc (hunk 1 above)
Comment (agrawroh) [reply test/extensions/filters/network/client_ssl_auth/client_ssl_auth_test.cc:170]: No, I am not seeing any issues. Just wanted to confirm once.

**Review by ggreenway — APPROVED** [review]

**Pull request merged**